crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -448,7 +466,251 @@ struct EthApiInner<Provider, Pool, Network> {
+#[serde(rename_all = "camelCase")]
+pub struct FeeHistoryCacheConfig {
+    /// Max number of blocks in cache.
+    ///
+    /// Default is 1024.
+    pub max_blocks: u64,
+}
+
+impl Default for FeeHistoryCacheConfig {
+    fn default() -> Self {
+        FeeHistoryCacheConfig { max_blocks: 1024 }
+    }
+}
+
+/// Wrapper struct for BTreeMap
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, FeeHistoryEntry>>>,
+    eth_cache: EthStateCache,
+}
+
+impl FeeHistoryCache {
+    /// Creates new FeeHistoryCache instance, initialize it with the mose recent data, set bounds
+    pub fn new(eth_cache: EthStateCache, config: FeeHistoryCacheConfig) -> Self {
+        let init_tree_map = BTreeMap::new();
+
+        let entries = Arc::new(tokio::sync::RwLock::new(init_tree_map));
+
+        let upper_bound = Arc::new(AtomicU64::new(0));
+        let lower_bound = Arc::new(AtomicU64::new(0));
+
+        FeeHistoryCache { config, entries, upper_bound, lower_bound, eth_cache }
+    }
+
+    /// Processing of the arriving blocks
+    pub async fn on_new_block<'a, I>(&self, blocks: I)
+    where
+        I: Iterator<Item = &'a Block>,
+    {
+        let mut entries = self.entries.write().await;
+
+        for block in blocks {
+            let mut fee_history_entry = FeeHistoryEntry::new(block, block.clone().seal_slow().hash);
+
+            let percentiles = predefined_percentiles();
+
+            fee_history_entry.rewards = calculate_reward_percentiles_for_block(
+                &percentiles,
+                &fee_history_entry,
+                self.eth_cache.clone(),
+            )
+            .await
+            .unwrap_or_default();
+
+            // calculate rewards
+            entries.insert(block.number, fee_history_entry);
+        }
+
+        while entries.len() > self.config.max_blocks as usize {
+            entries.pop_first();
+        }
+        if entries.len() == 0 {
+            self.upper_bound.store(0, SeqCst);
+            self.lower_bound.store(0, SeqCst);
+            return
+        }
+        let upper_bound = *entries.last_entry().expect("Contains at least one entry").key();
+        let lower_bound = *entries.first_entry().expect("Contains at least one entry").key();
+        self.upper_bound.store(upper_bound, SeqCst);
+        self.lower_bound.store(lower_bound, SeqCst);
+    }
+
+    /// Get UpperBound value for FeeHistoryCache
+    pub fn upper_bound(&self) -> u64 {
+        self.upper_bound.load(SeqCst)
+    }
+
+    /// Get LowerBound value for FeeHistoryCache
+    pub fn lower_bound(&self) -> u64 {
+        self.lower_bound.load(SeqCst)
+    }
+
+    /// Collect fee history for given range. It will try to use a cache to take the most recent
+    /// headers or if the range is out of caching config it will fallback to the database provider
```

Comment:
these docs are no longer accurate, there's no fallback here?
consider returning None if this is outside the range

Reply:
Return the empty vector if this is outside? right?